### PR TITLE
Refactor Txos

### PR DESCRIPTION
**Pull request by BrianCorbin**
Soundtrack of this PR: [link to song that really fits the mood of this PR]()

### Motivation

The goal is to decrease the complexity of and normalize the database. This tackles the txos table, and how we determine the status of a given TXO.

### In this PR
* Refactored txos table
* Refactored transaction_txos table into transaction_input_txos and transaction_output_txos
* Refactored how to determine status of txos

### Future Work
* continued refactoring of tables and code in preparation of support multiple token ids

### CODE REVIEW

File: full-service/src/db/account.rs [hunk 1]
```diff
@@ -492,16 +492,11 @@ impl AccountModel for Account {
 
         let mut accounts: Vec<Account> = Vec::<Account>::new();
 
-        if let Some(received_account_id_hex) = txo.received_account_id_hex {
+        if let Some(received_account_id_hex) = txo.account_id_hex {
```
Comment (Shramp) [inline full-service/src/db/account.rs:495]: small nit: rename `received_account_id_hex` to just `account_id_hex`. If we're not differentiating between received and minted, I think that not using the word `received` will avoid confusion for future devs

File: full-service/src/db/txo.rs [hunk 2]
```diff
@@ -639,23 +635,52 @@ impl TxoModel for Txo {
...
         token_id: Option<u64>,
         offset: Option<u64>,
         limit: Option<u64>,
         conn: &Conn,
     ) -> Result<Vec<Txo>, WalletDbError> {
-        use crate::db::schema::txos;
+        use crate::db::schema::{transaction_input_txos, transaction_logs, txos};
+
+        /*
+            SELECT * FROM txos
```
Comment (Shramp) [inline full-service/src/db/txo.rs:651]: Do we need to save this raw SQL comment?

File: full-service/src/db/transaction_log.rs [hunk 3]
```diff
@@ -202,30 +238,33 @@ impl TransactionLogModel for TransactionLog {
...
-            .load(conn)?;
+        use crate::db::schema::{transaction_input_txos, transaction_output_txos, txos};
 
-        let inputs = txos::table
-            .inner_join(
-                transaction_inputs::table.on(txos::txo_id_hex.eq(transaction_inputs::txo_id_hex)),
-            )
-            .filter(transaction_inputs::transaction_log_id.eq(&self.id))
+        let inputs: Vec<Txo> = txos::table
+            .inner_join(transaction_input_txos::table)
```
Comment (Shramp) [inline full-service/src/db/transaction_log.rs:244]: oh neat! you don't have to specify which columns to join on

File: full-service/src/db/txo.rs [hunk 4]
```diff
@@ -639,23 +635,52 @@ impl TxoModel for Txo {
...
+            transaction_logs::id
+                .is_null()
+                .or(transaction_logs::failed.eq(true))
+                .or(transaction_logs::id
+                    .is_not_null()
+                    .and(transaction_logs::submitted_block_index.is_null())),
+        );
+
+        query = query.filter(txos::received_block_index.is_not_null());
+        query = query.filter(txos::key_image.is_not_null());
```
Comment (Shramp) [inline full-service/src/db/txo.rs:685]: I think you need to add a filter for `txos::subaddress_index.is_not_null()` here

File: full-service/src/db/txo.rs [hunk 5]
```diff
@@ -15,22 +15,55 @@ use mc_transaction_core::{
...
+    Unspent,
+    Pending,
+    Spent,
+    Orphaned,
+}
+
+impl fmt::Display for TxoStatus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            TxoStatus::Unverified => write!(f, "unverified"),
```
Comment (Shramp) [inline full-service/src/db/txo.rs:46]: maybe not necessary for this PR, but I think a commented description for each txo status would be super helpful. Things to include: high-level description of what this status means, DB conditions for this status

File: full-service/src/service/txo.rs [hunk 6]
```diff
@@ -304,48 +297,48 @@ mod tests {
 
         // We should now have 3 txos - one pending, two minted (one of which will be
         // change)
-        let txos = service
-            .list_txos(&AccountID(alice.account_id_hex.clone()), None, None, None)
+        // let txos = service
```
Comment (Shramp) [inline full-service/src/service/txo.rs:300]: delete this comment?

File: full-service/src/service/txo.rs [hunk 7]
```diff
@@ -304,48 +297,48 @@ mod tests {
...
-            .cloned()
-            .filter(|txo| txo.minted_account_id_hex.is_some())
-            .collect();
-        assert_eq!(minted.len(), 2);
-        let minted_value_set = HashSet::from_iter(minted.iter().map(|m| m.value as u64));
-        assert!(minted_value_set.contains(&(58 * MOB - Mob::MINIMUM_FEE)));
-        assert!(minted_value_set.contains(&(42 * MOB)));
+        assert_eq!(pending[0].0.value, 100000000000000);
+
+        // let minted: Vec<Txo> = txos
```
Comment (Shramp) [inline full-service/src/service/txo.rs:324]: delete this comment?

**Review by Shramp — CHANGES_REQUESTED** [review]

File: full-service/src/db/txo.rs (hunk 4 above)
Comment (BrianCorbin) [reply full-service/src/db/txo.rs:685]: that's guaranteed by the key image not being null.

File: full-service/src/db/txo.rs (hunk 5 above)
Comment (BrianCorbin) [reply full-service/src/db/txo.rs:46]: will add! definitely useful and should be included :)

File: full-service/src/db/txo.rs (hunk 2 above)
Comment (BrianCorbin) [reply full-service/src/db/txo.rs:651]: I think it's nice to have the raw SQL query for reference, as reading the diesel stuff can be a little confusing if you're not familiar with it. Especially with filter chaining, the rules for that are very particular and diesel does some weird stuff sometimes that's unexpected.

File: full-service/src/db/account.rs (hunk 1 above)
Comment (BrianCorbin) [reply full-service/src/db/account.rs:495]: Hmm, I think that's the case currently. The txo table has a single account reference called "account_id_hex"

File: full-service/src/db/txo.rs (hunk 4 above)
Comment (Shramp) [reply full-service/src/db/txo.rs:694]: wait, then what was the point of this PR?

https://github.com/mobilecoinofficial/full-service/commit/f4f807ba7429521c8a3d021bf37a08eeb3a9a4b7

File: full-service/src/db/account.rs (hunk 1 above)
Comment (BrianCorbin) [reply full-service/src/db/account.rs:495]: ah I see! thanks, updated :)

File: full-service/src/db/txo.rs (hunk 4 above)
Comment (BrianCorbin) [reply full-service/src/db/txo.rs:694]: It was checking the wrong thing, it needed to check for key_image, not subaddress_index.

Comment (BrianCorbin) [reply full-service/src/db/txo.rs:694]: checking key_image or both key_image and subaddress index results in the same thing, but checking just subaddress_index will result in both unspent AND unverified txos

Comment (Shramp) [reply full-service/src/db/txo.rs:694]: ahhhh gotchya. Thanks

**Review by Shramp — APPROVED** [review]

**Pull request merged**